core/trino-main/src/main/java/io/trino/util/DateTimeUtils.java

```diff
@@ -67,9 +68,70 @@ public static int parseDate(String value)
         // - the required format is 'YYYY-MM-DD'
         // - all components should be unsigned numbers
         // https://github.com/trinodb/trino/issues/10677
+        Optional<Integer> days = tryParseDateYyyyMmDd(value);
+        if (days.isPresent()) {
+            return days.get();
+        }
         return toIntExact(TimeUnit.MILLISECONDS.toDays(DATE_FORMATTER.parseMillis(value)));
     }
 
+    /**
+     * Check if the date is in format yyyy-MM-dd, if yes, return the epoch days; otherwise
+     * return Optional.empty(), indicating the parse failed, and should fall back to generic
+     * parsing. Mixing the format checking and parsing in one pass for better performance.
+     * @param value date string to be parsed
+     * @return the number of days since 1970-01-01
+     */
+    public static Optional<Integer> tryParseDateYyyyMmDd(String value)
+    {
+        if (value.length() != 10) {
+            return Optional.empty();
+        }
+        // year
+        int year = value.charAt(0) - '0';
+        if (year < 0 || year > 9) {
+            return Optional.empty();
+        }
+        int curDigit = 0;
+        for (int i = 1; i < 4; i++) {
+            curDigit = value.charAt(i) - '0';
+            if (curDigit < 0 || curDigit > 9) {
+                return Optional.empty();
+            }
+            year = year * 10 + curDigit;
+        }
+        if (value.charAt(4) != '-') {
+            return Optional.empty();
+        }
+
+        // month
+        int month = value.charAt(5) - '0';
+        if (month < 0 || month > 9) {
+            return Optional.empty();
+        }
+        curDigit = value.charAt(6) - '0';
```

Comment:
I wouldn't reuse this local variable. You can instead have something like
```
int monthHigh = value.charAt(5) - '0';
...
int monthLow = value.charAt(6) - '0';
...
int dayHigh = value.charAt(8) - '0';
...
int dayLow = value.charAt(9) - '0';
```

Reply:
updated in 921b9c23e9894fdc57a581a29cb5d6aee89a99f9